### PR TITLE
Github: Support PR/issues with labels, full count

### LINKED ISSUE #420: Should GitHub issue badge including pull requests?
As you can see on the right side, GitHub put them in different pages and count individually, 
though we also get this info from GitHub api, but I wonder if this is a good flavor? Because it has different meaning with the count on GitHub project page.

Guys if you think they should be different, I can send a PR to adjust it, thanks!

### PR DESCRIPTION
**Pull request by dagwieers**
This PR includes the following:
- Simplified the code, but kept it full-featured
  - Use the issues API for all calls to Github
  - Got rid of the repos API for some of the queries
- It now displays more exact issue numbers (no longer indicates `30+`)
- It can now display the number of PRs by label too (not just issues only)
- Can display the used label name too for completeness

This fixes #839
This fixes #420
This implements #862 partly

image

### CODE REVIEW

**Comment by jhawkesworth** [comment]
nice!

**Comment by StoneCypher** [comment]
can the issues by-label include the label though?  without that it might quickly get confusing

**Comment by dagwieers** [comment]
@StoneCypher In our use-case, the context is very clear, so there is no need for that.
e.g. https://github.com/ansible/community/wiki/Windows:-action-list

But if you're going to add multiple of these on a single page, I see your point.
The question is how we are going to tell the code when the label should be shown.

Let me know what you like here. We could add -label to in there somewhere ?

**Comment by dagwieers** [comment]
@StoneCypher I now do this:

image

So basically, if you provide `-raw` it does not add the label, otherwise it will.
Is this OK for you ?

**Comment by StoneCypher** [comment]
ya 😄

**Comment by dagwieers** [comment]
@StoneCypher Enjoy !

PS Now I have to find out how to get this merged, because I would like to use it for real !

**Comment by StoneCypher** [comment]
Well, if I tag @espadrine , then you weren't the obnoxious one 😀

**Comment by dagwieers** [comment]
@StoneCypher Not sure if you had any summoning powers before, but they are surely failing now ;-)

**Comment by StoneCypher** [comment]
they did indeed

**Comment by StoneCypher** [comment]
@Daniel15 - any chance of looking at this?  This is rly valuable

**Comment by dagwieers** [comment]
@paulmelnikow Done. The very first existing Github test fails on my system (returns 404). So I had to comment it out during testing.

**Comment by StoneCypher** [comment]
@paulmelnikow (thank you)

**Comment by paulmelnikow** [comment]
Am I correct that `/github/issues` now returns only issues, and not issues + PRs?

File: service-tests/github.js [hunk 1]
```diff
@@ -205,3 +205,87 @@ t.create('hit counter for nonexistent repo')
...
+    name: Joi.equal('issues'),
+    value: Joi.string().regex(/^[0-9]+[kMGTPEZY]? open$/)
+  }));
+
+t.create('open issues (raw)')
+  .get('/issues-raw/badges/shields.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: Joi.equal('issues'),
+    value: Joi.string().regex(/^[0-9]+[kMGTPEZY]?$/)
+  }));
```
Comment (paulmelnikow) [inline service-tests/github.js]: What do you think about calling this `-minimalist` or `-clean`, or better yet, `/issues/minimalist`, rather than `-raw`? It seems to clarify that it's affecting the formatting, rather than the underlying number.

File: service-tests/github.js [hunk 2]
```diff
@@ -205,3 +205,87 @@ t.create('hit counter for nonexistent repo')
...
+    name: Joi.equal('pull requests'),
+    value: Joi.string().regex(/^[0-9]+[kMGTPEZY]?$/)
+  }));
+
+t.create('closed issues')
+  .get('/issues-closed/badges/shields.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: Joi.equal('closed issues'),
+    value: Joi.string().regex(/^[0-9]+[kMGTPEZY]? closed$/)
+  }));
```
Comment (paulmelnikow) [inline service-tests/github.js]: It seems to me that **closed** is not needed here, since the left side already says it's counting closed issues. Then we could omit `-closed-raw` as well.

File: service-tests/github.js [hunk 3]
```diff
@@ -205,3 +205,87 @@ t.create('hit counter for nonexistent repo')
...
+    name: Joi.equal('issues'),
+    value: Joi.string().regex(/^[0-9]+[kMGTPEZY]?$/)
+  }));
+
+t.create('open pull requests')
+  .get('/issues-pr/cdnjs/cdnjs.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: Joi.equal('pull requests'),
+    value: Joi.string().regex(/^[0-9]+[kMGTPEZY]? open$/)
+  }));
```
Comment (paulmelnikow) [inline service-tests/github.js]: Should this just be `/pull-requests` instead of `/issues-pr`?

Comment (paulmelnikow) [reply service-tests/github.js]: Ah, we need to support `/issues-pr` for backward compatibility. What do you think about supporting both?

**Comment by dagwieers** [comment]
> Am I correct that /github/issues now returns only issues, and not issues + PRs?

That is exactly as it was before AFAIK. You either have issues or PRs (and I don't know what the name is for both).

File: service-tests/github.js (hunk 1 above)
Comment (dagwieers) [reply service-tests/github.js]: I didn't introduce -raw, it was called like this before. I don't think we an break the existing interface.

File: service-tests/github.js (hunk 3 above)
Comment (dagwieers) [reply service-tests/github.js]: Yup, so I only simplified the implementation and made the numbers exact (rather than saying 30+ for everything larger than 30). And that's what we needed. The simplification was simple a result of the fact that it was needed for getting proper numbers.

File: service-tests/github.js (hunk 2 above)
Comment (dagwieers) [reply service-tests/github.js]: Rights, I think this was done because it's closer to what we do for open. Otherwise there's no difference anymore between `closed-raw` and `closed`.

Comment (StoneCypher) [reply service-tests/github.js]: almost my entire interest here is in closed issues (though i do not understand or care about the `-raw`) distinction

Comment (dagwieers) [reply service-tests/github.js]: Haha. Ok, I'll do what you suggested.

**Comment by dagwieers** [comment]
Done !

File: server.js [hunk 4]
```diff
@@ -3386,7 +3385,7 @@ cache(function(data, match, sendBadge, request) {
     try {
       var data = JSON.parse(buffer);
       var issues = data.total_count;
-      var rightText = isRaw? '': (isClosed? ' closed': ' open');
+      var rightText = isClosed? ' closed': (isRaw? '': ' open';
```
Comment (espadrine) [inline server.js]: Missing brace here. If you fix it quick, we might bring this to the server quite fast!

Comment (dagwieers) [reply server.js]: Right, should be fixed now. I also reorganized the tests, because there were some duplicates (after rebasing).

**Comment by dagwieers** [comment]
Latest result:
image

Unfortunately, with the used example for issues per label there are not issues. I was using a different repository in my previous examples.

**Comment by dagwieers** [comment]
Ready to merge ! (And then I can update our community wiki pages to get real numbers !)

**Comment by espadrine** [comment]
Merged and live.

**Comment by dagwieers** [comment]
@espadrine Thanks ! I was making some more small changes to the examples, but I'll add them to another PR. Nothing important.

**Comment by StoneCypher** [comment]
Working in production.  This used to say `issues 30+ closed` 😀 

screen shot 2017-09-24 at 2 37 36 pm

**Comment by paulmelnikow** [comment]
@dagwieers It's really cool to see this in use!

https://github.com/ansible/community/wiki/Ambassadors

**Comment by AraHaan** [comment]
@espadrine what about a badge for merged PR's only too?

**Comment by calebcartwright** [comment]
@AraHaan - thank you for looking back through the history to find things relevant to your inquiry, but a pull request that was closed more than 3 years isn't the best place for a new feature request. Please open a new issue if you'd like a new badge/new variant of a badge to increase the odds of someone potentially implementing it!

**Comment by AraHaan** [comment]
ok opened https://github.com/badges/shields/issues/6164
